homeassistant/components/nest.py

```diff
@@ -78,44 +103,61 @@ def nest_configuration_callback(data):
     )
 
 
-def setup_nest(hass, nest, config, pin=None):
+async def async_setup_nest(hass, nest, config, pin=None):
     """Set up the Nest devices."""
     if pin is not None:
         _LOGGER.debug("pin acquired, requesting access token")
         nest.request_token(pin)
 
     if nest.access_token is None:
         _LOGGER.debug("no access_token, requesting configuration")
-        request_configuration(nest, hass, config)
+        await async_request_configuration(nest, hass, config)
         return
 
     if 'nest' in _CONFIGURING:
         _LOGGER.debug("configuration done")
         configurator = hass.components.configurator
-        configurator.request_done(_CONFIGURING.pop('nest'))
+        configurator.async_request_done(_CONFIGURING.pop('nest'))
 
     _LOGGER.debug("proceeding with setup")
     conf = config[DOMAIN]
     hass.data[DATA_NEST] = NestDevice(hass, conf, nest)
 
-    _LOGGER.debug("proceeding with discovery")
-    discovery.load_platform(hass, 'climate', DOMAIN, {}, config)
-    discovery.load_platform(hass, 'camera', DOMAIN, {}, config)
+    _LOGGER.debug("proceeding with discovery -- climate")
+    await discovery.async_load_platform(hass, 'climate', DOMAIN, {}, config)
```

Comment:
Use `async_add_job` to schedule the discovery to avoid a deadlock where the component waits for the platform to be setup which waits for the component to be setup.

Reply:
Also use a for loop to that loads all the platforms. Iterate over component and platform config discovery info.

Reply:
Done